### PR TITLE
[Vulkan] Support uniform buffer object for passing many scalar arguments

### PR DESCRIPTION
**Pull request by masahi**
We are using push constants to pass scalar arguments to SPIR-V kernels. The vulkan spec ensures that the size of push constants storage is at least 128 bytes. Since we pass each scalar via 64 bit union, that means we can only pass 16 parameters via push constants. Unfortunately, for fused, dynamic input kernel, the number of `any_dim` and other params can easily go beyond that limit. See for example this crazy kernel that needs 20 scalars to be passed in https://gist.github.com/masahi/ce51c0d51c6115109203b3732f185aab

This PR enables passing parameters beyond push constants limit, using [uniform buffer object](https://vulkan.lunarg.com/doc/view/1.1.114.0/windows/tutorial/html/07-init_uniform_buffer.html). In particular, with this PR I was able to run PyTorch MaskRCNN end to end on Vulkan! 

Running NMS tests in `onnx/test_forward.py` serves as a test case.

There are some minor TODOs left, but I'd like to get some feedback. Supporting UBO requires making non trivial changes to both codegen and runtime, but hopefully they are straightforward.

@tqchen @tmoreau89 @jwfromm @ajtulloch  

TODO
- [x] Query platform specific push constant size limit rather than using a hard coded constant  
- [x] ~~Fix segfault on deleting UBO host buffer~~ (Seems I shouldn't allocate/deallocate the host buf explicitly, rather use vkMapMemory/vkUnMapMemory)
- [x] Add doc comments in ir_builder.h

### CODE REVIEW

**Comment by masahi** [comment]
@tqchen This is ready for review. 

On the codegen side, UBO is declared in a similar way as push constants, so I made `DeclareStorageVariable` function to declare both storage type.

On the runtime side, I introduced `CreateBuffer` function to declare both device storage buffers and uniform buffers. The change should be straightforward. 

One potentially controversial point could be the use of runtime function `MaxPushConstantsSize()` during codegen, to query the HW param and decide if we should use push constants or UBO.

**Comment by tqchen** [comment]
cc @ajtulloch @antinucleon please help to see if you can take a look

File: src/target/spirv/ir_builder.h [hunk 1]
```diff
@@ -555,6 +574,17 @@ class IRBuilder {
     val.flag = flag;
     return val;
   }
+
+  /*!
+   * \brief The common function to declare push constants and uniform buffe
```
Comment (tmoreau89) [inline src/target/spirv/ir_builder.h:579]: typo: buffer

**Review by tmoreau89 — APPROVED** [review]
Thank you @masahi for these changes, I left a comment on cosmetics

**Comment by tmoreau89** [comment]
Thanks you @masahi your PR has been merged

**Comment by tqchen** [comment]
Sorry that i didn't get a chance to review. Here are some items that needs to be fixed as followup PRs:

- Let us not use GetMaxPushConstantSize, as it is dynamic across devices, and the code won't behave correctly if we are doing cross compilation(where we compile on a machine that does not have vulkan), we should use a minimum TVM default (e.g. 128 bytes).
- I see a few places where delete is used, always prefer std::unique_ptr when possible instead of delete


@tmoreau89 given that you merged this PR, and merger usually needs to take responsibility in reviewing and fixing followup problems, can you send a followup PR to fix these two problems. Thank you!

**Comment by tqchen** [comment]
@tmoreau89 can you please send followup PRs to fix these problems?

**Comment by masahi** [comment]
@tqchen 

* I added one `delete` but I just follow the existing use of `delete` in https://github.com/apache/tvm/blob/5bc1cec4c4acf0a54889227c1d19a6b65b6803c2/src/runtime/vulkan/vulkan.cc#L239-L241 Since we need to manually delete or free associated vk related resources anyway, I don't mind this `delete` too much

*  Is there a way to make `GetMaxPushConstantSize` work for cross compilation use cases? Because for vulkan in particular, we definitely need to query HW related information like this at compile time. For example, the max thread block sizes, warp size, what extentions are available etc. I'm ok if we decide to use a fixed max push constant size, but for other HW params we need to have a solution for runtime querying.

**Comment by tqchen** [comment]
@masahi We can try to obtain these constant and put them into target. The main problem is the code as it is we are querying on the local machine but we need parameters on the remote. I think it is also important to consider de-coupling. e.g. The kernel should declare whether UBO or push constant is being used in its function meta-data, so we do not need to make dependency on the runtime parameters (e.g. the compiler decides whether to use UBO by arg size, and write that to an optional meta data of the shader to enable UBO). The runtime simply read that information and run dispatch, this will avoid inconsistency during compile time and runtime.

After taking a closer look, there are a few more things that needs to be fixed:

- Vulkan runtime have deferred mode and immediate mode that needs to be executed in separate ways(see the two places using push constants), we will need to do the same thing for UBO logic, right now seems that the code will work for immediate mode but not deferred mode
- Right now the UBO argument buffer simply re-allocates without deleting the original buffer, this could cause memory leaks. We will need to have a logic that re-allocates if necessary(also think about relation to synchronization in deferred case) and use the buffer if the size is large enough.

**Comment by masahi** [comment]
I confirmed that it also works for deferred mode (by always returning false from `UseImmediate()`). The relevent change is https://github.com/masahi/tvm/blob/1a3dbee99c9a2c362373707678d5657e59ea6827/src/runtime/vulkan/vulkan.cc#L1200-L1204  and https://github.com/masahi/tvm/blob/1a3dbee99c9a2c362373707678d5657e59ea6827/src/runtime/vulkan/vulkan.cc#L1215  

Need to look at the reallocate problem

**Comment by tqchen** [comment]
Sorry I was wrong about the re-allocation, seems you are allocating a single UBO per-kernel, so it might be fine. My original thought was along the line of single UBO per runtime.

**Comment by masahi** [comment]
Yes, one UBO per one pipeline https://github.com/masahi/tvm/blob/1a3dbee99c9a2c362373707678d5657e59ea6827/src/runtime/vulkan/vulkan.cc#L108

**Comment by tqchen** [comment]
To followup on the deffered execution case:

Because memcpy happens here: https://github.com/masahi/tvm/blob/1a3dbee99c9a2c362373707678d5657e59ea6827/src/runtime/vulkan/vulkan.cc#L1150

Imagine there are two consecutive launch of the same kernel A(uses UBO) that are deferred. Then the memcpy of the second kernel launch would override the value of the first launch(both of which hasn't yet been launched).

**Comment by tqchen** [comment]
To followup on the per pipeline UBO. This was fine for single threaded case, but can also be problematic under a multi-threaded setting when multiple threads are launching the same kernel A. 

To make the runtime thread-safe, we normally needs to divide the data structure into constants(e.g. pipeline) and runtime structure(e.g. staging buffer, streams). The runtime structure part belongs to VulkanThreadEntry that comes with a thread-local copy to avoid threading issue. So we would want to do that, and uses a similar logic as StagingBuffer to create staging buffer for UBO

**Comment by tqchen** [comment]
Thanks @masahi for great discussions. It seems to me that we would need to spend a bit more time reviewing and discussing the PR. Given it is already merged, we could consider:

- A0: Revert the PR, and bring another one, do another round of thorough reviews there.
- A1: Bring a fix PR as soon as possible, in this case the original merger(@tmoreau89) should make sure all the points are being addressed.

To summarize the items that needs to be addressed:

- P0: Add the use of UBO as meta-data to shader, so we won't have compiler runtime in-consistentcy problem
- P1: Solve the deferred execution case when a UBO buffer can be overriden by the followup kernel calls(need to do the copy in deferred stream as well).
- P2: Solve the multi-thread safety issue when multiple threads calls into UBO at the same time.

**Comment by masahi** [comment]
ok sent the revert https://github.com/apache/tvm/pull/7821

**Comment by tmoreau89** [comment]
@masahi @tqchen Thanks for the comments, it appears as though the merge might have been to hasty. I'm good with the decision to go with A0; thanks @masahi for issuing the revert.
